tsdb/chunks/chunk_write_queue.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package chunks
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+)
+
+type chunkWriteJob struct {
+	cutFile   bool
+	seriesRef HeadSeriesRef
+	mint      int64
+	maxt      int64
+	chk       chunkenc.Chunk
+	ref       ChunkDiskMapperRef
+	callback  func(error)
+}
+
+var (
+	queueOperationAdd      = "add"
+	queueOperationGet      = "get"
+	queueOperationComplete = "complete"
+	queueOperations        = []string{queueOperationAdd, queueOperationGet, queueOperationComplete}
+)
+
+// chunkWriteQueue is a queue for writing chunks to disk in a non-blocking fashion.
+// Chunks that shall be written get added to the queue, which is consumed asynchronously.
+// Adding jobs to the job is non-blocking as long as the queue isn't full.
+type chunkWriteQueue struct {
+	size  int
+	jobCh chan chunkWriteJob
+
+	chunkRefMapMtx       sync.RWMutex
+	chunkRefMap          map[ChunkDiskMapperRef]chunkenc.Chunk
+	chunkRefMapOversized bool // indicates whether more than <size> chunks were put into the chunkRefMap.
+
+	isRunningMtx sync.RWMutex
+	isRunning    bool
+
+	workerWg sync.WaitGroup
+
+	writeChunk writeChunkF
+
+	operationsMetric *prometheus.CounterVec
+}
+
+// writeChunkF is a function which writes chunks, it is dynamic to allow mocking in tests.
+type writeChunkF func(HeadSeriesRef, int64, int64, chunkenc.Chunk, ChunkDiskMapperRef, bool) error
+
+func newChunkWriteQueue(reg prometheus.Registerer, size int, writeChunk writeChunkF) *chunkWriteQueue {
+	q := &chunkWriteQueue{
+		size:        size,
+		jobCh:       make(chan chunkWriteJob, size),
+		chunkRefMap: make(map[ChunkDiskMapperRef]chunkenc.Chunk, size),
+		writeChunk:  writeChunk,
+
+		operationsMetric: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "prometheus_tsdb_chunk_write_queue_operations_total",
+				Help: "Number of operations on the chunk_write_queue.",
+			},
+			[]string{"operation"},
+		),
+	}
+
+	if reg != nil {
+		reg.MustRegister(q.operationsMetric)
+
+		// Initialize series for all the possible labels.
+		for _, op := range queueOperations {
+			q.operationsMetric.WithLabelValues(op).Add(0)
```

Comment:
Nitpick: afaik there's no need to `.Add(0)`.

Reply:
It's not necessary for the metric to work, correct. The purpose of adding `0` here is just to initialize the metric when the chunk write queue gets started.

Reply:
Yes, I meant that it's not needed to `.Add(0)` to initialise the metric, `WithLabelValues()` already initialises it to 0.

See this example, although golang playground can't run it: https://go.dev/play/p/22GMtzeal5J

Reply:
oh ic, thanks! Will update

Reply:
done here 3f1544c2e154c93ce60ad105f4d2b9c4e84a7b90